open-machine-learning-jupyter-book/ml-advanced/clustering/introduction-to-clustering.md

```diff
@@ -0,0 +1,323 @@
+    df = df[(df['popularity'] > 0)]
+    top = df['artist_top_genre'].value_counts()
+    plt.figure(figsize=(10,7))
+    sns.barplot(x=top.index,y=top.values)
+    plt.xticks(rotation=45)
+    plt.title('Top genres',color = 'blue')
+    ```
+
+1. Do a quick test to see if the data correlates in any particularly strong way:
+
+    ```python
+    corrmat = df.corr()
+    f, ax = plt.subplots(figsize=(12, 9))
+    sns.heatmap(corrmat, vmax=.8, square=True)
+    ```
+
+    ![correlations](../../../images/clustering/correlation.png)
+
+    The only strong correlation is between `energy` and `loudness`, which is not too surprising, given that loud music is usually pretty energetic. Otherwise, the correlations are relatively weak. It will be interesting to see what a clustering algorithm can make of this data.
+
+    > 🎓 Note that correlation does not imply causation! We have proof of correlation but no proof of causation. An amusing web site has some visuals that emphasize this point.
+
+Is there any convergence in this dataset around a song's perceived popularity and danceability? A FacetGrid shows that there are concentric circles that line up, regardless of genre. Could it be that Nigerian tastes converge at a certain level of danceability for this genre?  
+
+✅ Try different datapoints (energy, loudness, speechiness) and more or different musical genres. What can you discover? Take a look at the `df.describe()` table to see the general spread of the data points.
+
+### Exercise - data distribution
+
+Are these three genres significantly different in the perception of their danceability, based on their popularity?
+
+1. Examine our top three genres data distribution for popularity and danceability along a given x and y axis.
+
+    ```python
+    sns.set_theme(style="ticks")
+    
+    g = sns.jointplot(
+        data=df,
+        x="popularity", y="danceability", hue="artist_top_genre",
+        kind="kde",
+    )
+    ```
+
+    You can discover concentric circles around a general point of convergence, showing the distribution of points.
+
+    > 🎓 Note that this example uses a KDE (Kernel Density Estimate) graph that represents the data using a continuous probability density curve. This allows us to interpret data when working with multiple distributions.
+
+    In general, the three genres align loosely in terms of their popularity and danceability. Determining clusters in this loosely-aligned data will be a challenge:
+
+    ![distribution](../../../images/clustering/distribution.png)
+
+1. Create a scatter plot:
+
+    ```python
+    sns.FacetGrid(df, hue="artist_top_genre", size=5) \
+       .map(plt.scatter, "popularity", "danceability") \
+       .add_legend()
+    ```
+
+    A scatterplot of the same axes shows a similar pattern of convergence
+
+    ![Facetgrid](../../../images/clustering/facetgrid.png)
```

Comment:
If you make the above code as {code-cell}, it will be executed and output this image for you. You should remove such images.